tiledb/api/src/query/subarray.rs

```diff
@@ -359,6 +359,38 @@ where
         Ok(b)
     }
 
+    pub fn set_subarray<IntoRange: Into<Range>>(
+        self,
+        range: IntoRange,
+    ) -> TileDBResult<Self> {
+        let c_subarray = *self.raw;
+
+        let range = range.into();
+        // TODO: should we be checking dimensionality here?
+
+        match range {
+            Range::Single(range) => {
+                single_value_range_go!(range, _DT, start, end, {
+                    let start = start.to_le_bytes();
+                    let _end = end.to_le_bytes();
+
+                    self.query.base().capi_call(|ctx| unsafe {
+                        // TODO: is this correct?
+                        ffi::tiledb_subarray_set_subarray(
+                            ctx,
+                            c_subarray,
+                            start.as_ptr() as *const std::ffi::c_void,
+                        )
+                    })?;
+                })
+            }
+            Range::Multi(_) => unreachable!("Unreachable (TODO)"),
```

Comment:
Rust has `unimplemented!()` and `todo!()` as well which are quite nice.

Reply:
I am following the example from here: https://github.com/TileDB-Inc/tiledb-rs/blob/bcf910a94b32202bf8e560bf3beb424ed8012677/tiledb/api/src/query/subarray.rs#L264

This is more of a question. Should I be worrying about Range::Multi and Range::Var at this point. I'm not clear about the intent behind Single, Multi and Var.

Reply:
I don't think so.
`Range::Multi` AFAIK is not used at all in practice, maybe it will be someday but isn't there yet.
`Range::Var` can show up on string-typed dimensions for sparse arrays, but for those one would have to use a different function to set the range.


Aside:

Attributes and dimensions both have a "cell val num" which is the number of values of the type that occur per cell in the array.
- a value of 1 is kind of what you would expect
- a fixed value N of greater than 1 means that exactly N values appear per cell.  This is not common, unclear if it is ever used at all.  If it is, surely it is only small numbers.  And it definitely never appears on a dimension.
- `CellValNum::Var` is what it looks like in `tiledb-rs`, in core this is the constant `u32::MAX`, but either way this value means that the number of values per cell is variable.  To query these in C you need an additional offsets column - in `tiledb-rs` it can be done that way but there are also iterators to abstract it away.

Dimensions of dense arrays always have cell val num 1.  Dimensions of sparse arrays do too, *except* `StringAscii` typed dimensions can have `CellValNum::Var`.

Reply:
Thanks for explaining, will update based on comments above.

Reply:
To be clear, do you propose both be unimplemented?